open_spiel/games/dark_chess.cc

```diff
@@ -87,47 +87,156 @@ std::array<bool, chess::kMaxBoardSize * chess::kMaxBoardSize> ComputeObservabili
+    std::string result;
+
+    if (iig_obs_type_.perfect_recall) {
+      if (iig_obs_type_.public_info && iig_obs_type_.private_info == PrivateInfoType::kSinglePlayer) {
+        return state.aohs_[player].ToString();
+      } else {
+        SpielFatalError("Dark chess observer string with perfect recall implemented only for the default info state observation type");
+      }
+    }
+
+    if (iig_obs_type_.public_info && iig_obs_type_.private_info == PrivateInfoType::kSinglePlayer) {
+      chess::Color color = chess::PlayerToColor(player);
+      auto observability_table = ComputeObservabilityTable(state.Board(), color);
+      return state.Board().ToDarkFEN(observability_table, color);
+    }
+    else {
+      SpielFatalError("Dark chess observer string with imperfect recall implemented only for the default observation type");
+    }
+
   }
-}
 
-// Adds a uniform scalar plane scaled with min and max.
-template <typename T>
-void AddScalarPlane(T val, T min, T max,
-                    absl::Span<float>::iterator& value_it) {
-  double normalized_val = static_cast<double>(val - min) / (max - min);
-  for (int i = 0; i < chess::kMaxBoardSize * chess::kMaxBoardSize; ++i)
-    *value_it++ = normalized_val;
-}
+ private:
+
+  void WritePieces(chess::Color color, chess::PieceType piece_type,
+                         const chess::ChessBoard &board,
+                         std::array<bool, chess::kMaxBoardSize * chess::kMaxBoardSize> &observability_table,
+                         Allocator* allocator) const {
+
+    std::string typeString = color == chess::Color::kEmpty ? "empty" : chess::PieceTypeToString(piece_type,color == chess::Color::kWhite);
+
+    auto out = allocator->Get(typeString + "_pieces", {chess::kMaxBoardSize, chess::kMaxBoardSize});
+    for (int8_t y = 0; y < chess::kMaxBoardSize; ++y) {
+      for (int8_t x = 0; x < chess::kMaxBoardSize; ++x) {
+        auto square = chess::Square{x, y};
+        chess::Piece piece_on_board = board.at(square);
+        out.at(x ,y) =
+            (piece_on_board.color == color && piece_on_board.type == piece_type && observability_table[chess::SquareToIndex(square, board.BoardSize())]
+             ? 1.0f
+             : 0.0f);
+      }
+    }
+  }
+
+  void WriteUnknownSquares(const chess::ChessBoard &board,
+                           std::array<bool, chess::kMaxBoardSize * chess::kMaxBoardSize> &observability_table,
+                           Allocator *allocator) const {
+
+    auto out = allocator->Get("unknown_squares", {chess::kMaxBoardSize, chess::kMaxBoardSize});
+    for (int8_t y = 0; y < chess::kMaxBoardSize; ++y) {
+      for (int8_t x = 0; x < chess::kMaxBoardSize; ++x) {
+        auto square = chess::Square{x, y};
+        out.at(x, y) = observability_table[chess::SquareToIndex(square, board.BoardSize())] ? 0.0 : 1.0;
+      }
+    }
+  }
+
+  template <typename T>
+  void WriteScalar(T val, T min, T max, Allocator *allocator, const std::string &name) const {
```

Comment:
Not all scalars need to occupy the whole board - use this function only for those that do need to use the board.